Feat/bulk-delete-entity

Joao-vi opened the pull request:
fixes #8628

PR checklist:
- [ ] Update READ.me ?
- [ ] Update API documentation ?

QA checklist:
- [ ] Smoke test the functionality described in the issue
- [ ] Test for side effects
- [ ] UI responsiveness
- [ ] Cross browser testing
- [ ] Code review

Joao-vi committed:
wip use case

Joao-vi committed:
add feature flag for bulk delete entity use case

Joao-vi committed:
add bulk delete by shared id on elastic search

Joao-vi committed:
add bulk delete entity use case

Joao-vi committed:
Merge branch 'production' into feat/bulk-delete-entity

daneryl commented on app/api/core/application/BulkDeleteEntity.ts:
```diff
@@ -0,0 +1,43 @@
+      this.jobsDispatcher.dispatchMany(async dispatch =>
+        dispatch(BatchDeleteEntityJob, { sharedIds: chunk })
+      )
+    );
```
In order to take advantage here of the dispatchMany it should be inverted:

```ts
    this.jobsDispatcher.dispatchMany(
      async dispatch =>
        await ArrayUtils.sequentialFor(chunks, async chunk =>
          dispatch(BatchDeleteEntityJob, { sharedIds: chunk })
        )
    );
```

otherwise its the same as using the normal ```jobsDispatcher.dispatch()```
also probably you do not need ArrayUtils here, just forEach or for loop

Joao-vi committed:
Merge branch 'production' into feat/bulk-delete-entity

Joao-vi committed:
refactor: enhance bulk deletion process with transaction management and error handling

Joao-vi committed:
feat: Implement BatchDeleteEntity use case and related infrastructure

- Added BatchDeleteEntityUseCase for handling bulk deletion of entities.
- Introduced EntityPermissionChecker and MongoEntityPermissionChecker for permission validation.
- Updated BulkDeleteEntityUseCase to utilize the new permission checker.
- Created BatchDeleteEntityJob for processing deletion jobs in the background.
- Enhanced MultiLanguageEntityDataSource with bulkDelete and getAllBySharedId methods.
- Added tests for BatchDeleteEntityUseCase and BulkDeleteEntityUseCase to ensure correct functionality and permission handling.
- Updated User model to include a static method for creating user instances from props.

Joao-vi committed:
fine tunning bulk delete use case

Joao-vi committed:
wip

Joao-vi committed:
refactor: remove RequestBulkDeleteEntity and related tests, introduce BulkCleanupEntity

- Deleted RequestBulkDeleteEntity.spec.ts and associated use case.
- Introduced BulkCleanupEntityUseCase and its factory.
- Updated EntityPermissionChecker to return templateId along with sharedId.
- Modified MongoEntityPermissionChecker to handle new output structure.
- Created BulkCleanupEntityJob to handle bulk cleanup operations.
- Updated BulkDeleteEntityController to use new BulkCleanupEntityUseCase.
- Implemented deleteReferencesToSharedIds method in MultiLanguageEntityDataSource.
- Adjusted relationships data source to support bulk deletion by sharedId.
- Enhanced search functionality to refresh after bulk operations.
- Updated queue registry to register new BulkCleanupEntityJob.

Joao-vi commented on app/api/core/application/BulkCleanupEntity.ts:
```diff
@@ -0,0 +1,50 @@
+
+    await this.transactionManager.run(async () => {
+      await this.deps.relationshipsDS.bulkDeleteBySharedId(sharedIds);
+      await this.deps.entitiesDS.deleteReferencesToSharedIds(deleteEntities);
```
`await this.deps.entitiesDS.deleteReferencesToSharedIds(deleteEntities)`

This operation turns out to be a big problem: 

Amount of documents involved here are not under of our control, a tenant might have entity deleted that might be referenced in thousand of thousands of other entities. 

But here is even worse because is a bulk operation, image 100 deleted entities that might be connected to infinite of documents that needs to be cleaned up (remove the deleted sharedId from its metadata values).

I tested locally, deleting 930 entities, first part of the process went smoothly (`BulkDeleteEntityUseCase`), now while running this use case (`BulkCleanupEntityUseCase`) for the 10 jobs created, 9 were chunks of 100 entities, the only one that was able to finish was of the chunk of 30 entities (still this got 4 retries). 

All errors messages was regarding transactions timeout.

@daneryl

Joao-vi replied:
Big win! 

I was able to improve this a lot! We went from a state where Jobs with 100 chunks were not even be able to be executed to being now executed in about 1.8~2s

The biggest bottle neck was quering all affected entities so we can dispatch reindex and to update on mongo. In order to do that we were scanning the whole entities collection. 

I am now, first, filtering entities that have metadata that might have a reference for the entity as a value,  this being `select`, `multiselect` and `relationship` 

@daneryl

Joao-vi committed:
Merge branch 'production' of github.com:huridocs/uwazi into feat/bulk-delete-entity

Joao-vi committed:
Merge branch 'production' of github.com:huridocs/uwazi into feat/bulk-delete-entity

Joao-vi committed:
Enhance bulk deletion tests and improve shared ID reference checks in MongoMultiLanguageEntityDataSource

Joao-vi committed:
improve BulkCleanupEntityUseCase performance

Joao-vi committed:
improve query on entities collection by providing a default language

Joao-vi committed:
Merge branch 'production' of github.com:huridocs/uwazi into feat/bulk-delete-entity

Joao-vi committed:
refactor: disable v2BulkDeleteEntity feature and enhance BulkCleanupEntityUseCase with file deletion logic

Joao-vi committed:
feat: add bulk delete validation schema and update controller logic

Joao-vi committed:
Merge branch 'production' of github.com:huridocs/uwazi into feat/bulk-delete-entity

Joao-vi committed:
re throw error

Joao-vi committed:
fix: correct import path for relationships module
